Bottom: 6bbb92f637582964eea7a287f91549c84da9571a
Top:    f808773d213c6b0b6ecbaff3f68d99d199d218b8
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-08-07 21:30:48 -0700

support for debug prints

---

diff --git a/src/preload.c b/src/preload.c
index 53f8543..fdb25ac 100644
--- a/src/preload.c
+++ b/src/preload.c
@@ -48,6 +48,8 @@
 #include <netinet/tcp.h>
 #include <unistd.h>
 #include <semaphore.h>
+#include <stdio.h>
+#include <sys/syscall.h>
 
 #include <rdma/rdma_cma.h>
 #include <rdma/rdma_verbs.h>
@@ -98,6 +100,8 @@ static int rq_size;
 static int sq_inline;
 static int fork_support;
 
+static FILE *fout;
+
 enum fd_type {
 	fd_normal,
 	fd_rsocket,
@@ -286,6 +290,7 @@ static void init_preload(void)
 	rs.getsockopt = dlsym(RTLD_DEFAULT, "rgetsockopt");
 	rs.fcntl = dlsym(RTLD_DEFAULT, "rfcntl");
 
+	fout = fopen("rs-out.txt", "w+");
 	getenv_options();
 	init = 1;
 out:
@@ -391,6 +396,8 @@ int socket(int domain, int type, int protocol)
 	if (index < 0)
 		return index;
 
+	fprintf(fout, "%d socket %d\n", syscall(SYS_gettid), index);
+	fflush(fout);
 	recursive = 1;
 	ret = rsocket(domain, type, protocol);
 	recursive = 0;
@@ -405,6 +412,9 @@ int socket(int domain, int type, int protocol)
 			fd_store(index, ret, fd_rsocket);
 			set_rsocket_options(ret);
 		}
+		fprintf(fout, "%d socket %d real fd %d type %d\n",
+				syscall(SYS_gettid), index, ret, fd_gett(index));
+		fflush(fout);
 		return index;
 	}
 	fd_close(index, &ret);
@@ -415,6 +425,8 @@ real:
 int bind(int socket, const struct sockaddr *addr, socklen_t addrlen)
 {
 	int fd;
+	fprintf(fout, "%d bind %d\n", syscall(SYS_gettid), socket);
+	fflush(fout);
 	return (fd_get(socket, &fd) == fd_rsocket) ?
 		rbind(fd, addr, addrlen) : real.bind(fd, addr, addrlen);
 }
@@ -422,6 +434,8 @@ int bind(int socket, const struct sockaddr *addr, socklen_t addrlen)
 int listen(int socket, int backlog)
 {
 	int fd;
+	fprintf(fout, "%d listen %d\n", syscall(SYS_gettid), socket);
+	fflush(fout);
 	return (fd_get(socket, &fd) == fd_rsocket) ?
 		rlisten(fd, backlog) : real.listen(fd, backlog);
 }
@@ -432,6 +446,8 @@ int accept(int socket, struct sockaddr *addr, socklen_t *addrlen)
 	enum fd_type type;
 
 	type = fd_get(socket, &fd);
+	fprintf(fout, "%d accept %d\n", syscall(SYS_gettid), socket);
+	fflush(fout);
 	if (type == fd_rsocket || type == fd_fork) {
 		index = fd_open();
 		if (index < 0)
@@ -444,6 +460,9 @@ int accept(int socket, struct sockaddr *addr, socklen_t *addrlen)
 			return ret;
 		}
 
+		fprintf(fout, "%d accept %d new fd %d new real fd %d\n",
+			syscall(SYS_gettid), socket, index, ret);
+		fflush(fout);
 		fd_store(index, ret, type);
 		return index;
 	} else {
@@ -465,6 +484,8 @@ static int fork_active(int socket, const struct sockaddr *addr, socklen_t addrle
 	long flags;
 
 	fd = fd_getd(socket);
+	fprintf(fout, "%d fork_active - %d fd %d\n", syscall(SYS_gettid), socket, fd);
+	fflush(fout);
 	flags = real.fcntl(fd, F_GETFL);
 	real.fcntl(fd, F_SETFL, 0);
 	ret = real.connect(fd, addr, addrlen);
@@ -495,6 +516,8 @@ static void fork_passive(int socket)
 	uint32_t msg;
 
 	fd_get(socket, &sfd);
+	fprintf(fout, "%d fork_passive %d real fd %d\n", syscall(SYS_gettid), socket, sfd);
+	fflush(fout);
 
 	len = sizeof sin6;
 	ret = real.getsockname(sfd, (struct sockaddr *) &sin6, &len);
@@ -527,11 +550,17 @@ static void fork_passive(int socket)
 	ret = rlisten(lfd, 1);
 	if (ret)
 		goto lclose;
+	fprintf(fout, "%d fork_passive %d, listening for fork\n", syscall(SYS_gettid), socket);
+	fflush(fout);
 
 	msg = 0;
 	len = real.write(sfd, &msg, sizeof msg);
 	if (len != sizeof msg)
+	{
+		fprintf(fout, "%d fork_passive - %d write failed\n", syscall(SYS_gettid), socket);
+		fflush(fout);
 		goto lclose;
+	}
 
 	dfd = raccept(lfd, NULL, NULL);
 	if (dfd < 0) {
@@ -547,6 +576,9 @@ static void fork_passive(int socket)
 	real.shutdown(sfd, SHUT_RDWR);
 	real.close(sfd);
 	fd_store(socket, dfd, fd_rsocket);
+	fprintf(fout, "%d fork_passive success %d new fd %d old fd %d\n",
+		syscall(SYS_gettid), socket, dfd, sfd);
+	fflush(fout);
 
 lclose:
 	rclose(lfd);
@@ -628,6 +660,8 @@ ssize_t read(int socket, void *buf, size_t count)
 {
 	int fd;
 	init_preload();
+	fprintf(fout, "%d read %d read fd %d\n", syscall(SYS_gettid), socket, fd_getd(socket));
+	fflush(fout);
 	return (fd_fork_get(socket, &fd) == fd_rsocket) ?
 		rread(fd, buf, count) : real.read(fd, buf, count);
 }
@@ -820,6 +854,8 @@ int close(int socket)
 	int ret;
 
 	init_preload();
+	fprintf(fout, "%d close %d real fd %d\n", syscall(SYS_gettid), socket, fd_getd(socket));
+	fflush(fout);
 	fdi = idm_lookup(&idm, socket);
 	if (!fdi)
 		return real.close(socket);
@@ -924,6 +960,8 @@ int dup2(int oldfd, int newfd)
 	int ret;
 
 	init_preload();
+	fprintf(fout, "%d dup2 %d -> %d\n", syscall(SYS_gettid), oldfd, newfd);
+	fflush(fout);
 	oldfdi = idm_lookup(&idm, oldfd);
 	if (oldfdi && oldfdi->type == fd_fork)
 		fork_passive(oldfd);
@@ -961,6 +999,9 @@ int dup2(int oldfd, int newfd)
 	atomic_init(&newfdi->refcnt);
 	atomic_set(&newfdi->refcnt, 1);
 	atomic_inc(&oldfdi->refcnt);
+	fprintf(fout, "%d dup2 %d -> %d, real fd %d type %d\n",
+		syscall(SYS_gettid), oldfd, newfd, newfdi->fd, newfdi->type);
+	fflush(fout);
 	return newfd;
 }
 
@@ -968,12 +1009,18 @@ int __fxstat64(int ver, int socket, struct stat64 *buf)
 {
 	int fd, ret;
 
+	init_preload();
+	fprintf(fout, "%d fxstat64 socket %d - fd %d type %d\n",
+			syscall(SYS_gettid), socket, fd_getd(socket), fd_gett(socket));
 	if (fd_get(socket, &fd) == fd_rsocket) {
 		ret = real.fxstat64(ver, socket, buf);
+		fprintf(fout, "%d fxstat64 - rsocket %d\n", syscall(SYS_gettid), ret);
 		if (!ret)
 			buf->st_mode = (buf->st_mode & ~S_IFMT) | __S_IFSOCK;
 	} else {
 		ret = real.fxstat64(ver, fd, buf);
+		fprintf(fout, "%d fxstat64 - normal %d\n", syscall(SYS_gettid), ret);
 	}
+	fflush(fout);
 	return ret;
 }
